File: driftwatch/backend/internal/alerts/discord.go
package alerts

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/YOURUSERNAME/driftwatch/internal/agent"
	"github.com/YOURUSERNAME/driftwatch/internal/gemini"
)

const cooldownPeriod = 30 * time.Minute

type Client struct {
	webhookURL string
	http       *http.Client
}

func NewClient(webhookURL string) *Client {
	return &Client{
		webhookURL: webhookURL,
		http:       &http.Client{Timeout: 10 * time.Second},
	}
}

// Discord webhook payload shapes.
type discordPayload struct {
	Embeds []discordEmbed `json:"embeds"`
}

type discordEmbed struct {
	Title     string         `json:"title"`
	Color     int            `json:"color"`
	Fields    []discordField `json:"fields"`
	Footer    discordFooter  `json:"footer"`
	Timestamp string         `json:"timestamp"`
}

type discordField struct {
	Name   string `json:"name"`
	Value  string `json:"value"`
	Inline bool   `json:"inline,omitempty"`
}

type discordFooter struct {
	Text string `json:"text"`
}

func colorForSeverity(sev string) int {
	switch sev {
	case agent.SeverityCritical:
		return 15158332 // red
	case agent.SeverityWarning:
		return 16776960 // yellow
	case agent.SeverityInfo:
		return 3447003 // blue
	default:
		return 3447003
	}
}

func uniqueContainerCount(drifts []agent.DriftEvent) int {
	seen := make(map[string]struct{}, len(drifts))
	for _, d := range drifts {
		seen[d.ContainerName] = struct{}{}
	}
	return len(seen)
}

func (c *Client) SendDriftAlert(projectName string, result *gemini.AnalysisResult, drifts []agent.DriftEvent) error {
	if result == nil {
		return errors.New("alerts: AnalysisResult is nil")
	}

	now := time.Now().UTC()
	timestamp := now.Format(time.RFC3339)

	embed := discordEmbed{
		Title: fmt.Sprintf("⚠️ Drift Detected — %s", projectName),
		Color: colorForSeverity(result.Severity),
		Fields: []discordField{
			{Name: "Severity", Value: result.Severity, Inline: true},
			{Name: "Containers Affected", Value: fmt.Sprintf("%d", uniqueContainerCount(drifts)), Inline: true},
			{Name: "Summary", Value: result.Summary},
			{Name: "Fix Command", Value: fmt.Sprintf("```bash\n%s\n```", result.FixCommand)},
			{Name: "Explanation", Value: result.Explanation},
		},
		Footer:    discordFooter{Text: fmt.Sprintf("DriftWatch • detected at %s", timestamp)},
		Timestamp: timestamp,
	}

	body, err := json.Marshal(discordPayload{Embeds: []discordEmbed{embed}})
	if err != nil {
		return fmt.Errorf("marshal discord payload: %w", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.webhookURL, bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("build discord request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.http.Do(req)
	if err != nil {
		return fmt.Errorf("discord post: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		respBody, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("discord webhook returned %d: %s", resp.StatusCode, string(respBody))
	}

	return nil
}

// ShouldAlert returns true if it's been at least 30 minutes since the last
// alert for this project (or there's no prior alert recorded).
//
// projectID is accepted for forward compatibility with per-project state,
// but the current implementation is stateless and relies on the caller to
// pass the last-alerted timestamp from the DB.
func (c *Client) ShouldAlert(projectID string, lastAlertedAt *time.Time) bool {
	_ = projectID
	if lastAlertedAt == nil {
		return true
	}
	return time.Since(*lastAlertedAt) >= cooldownPeriod
}

File: driftwatch/backend/src/services/discord.ts
import { Severity } from '../types.js';
import type { DriftEvent } from '../types.js';
import type { AnalysisResult } from './gemini.js';

const COOLDOWN_MS = 30 * 60 * 1000;

export interface DiscordField {
	name: string;
	value: string;
	inline?: boolean;
}

export interface DiscordEmbed {
	title: string;
	color: number;
	fields: DiscordField[];
	footer: { text: string };
	timestamp: string;
}

export interface DiscordPayload {
	embeds: DiscordEmbed[];
}

export function colorForSeverity(sev: string): number {
	switch (sev) {
		case Severity.Critical:
			return 15158332; // red
		case Severity.Warning:
			return 16776960; // yellow
		case Severity.Info:
			return 3447003; // blue
		default:
			return 3447003;
	}
}

function uniqueContainerCount(drifts: DriftEvent[]): number {
	return new Set((drifts ?? []).map((d) => d.containerName)).size;
}

export class DiscordClient {
	private readonly webhookUrl: string;
	private readonly timeoutMs: number;
	private readonly fetchImpl: typeof fetch;

	constructor(
		webhookUrl: string,
		opts: { timeoutMs?: number; fetchImpl?: typeof fetch } = {}
	) {
		this.webhookUrl = webhookUrl;
		this.timeoutMs = opts.timeoutMs ?? 10_000;
		this.fetchImpl = opts.fetchImpl ?? fetch;
	}

	/**
	 * Posts to the client's configured webhook URL. Kept for backward
	 * compatibility; multi-user callers use sendDriftAlertTo with the project's
	 * own webhook.
	 */
	async sendDriftAlert(
		projectName: string,
		result: AnalysisResult | null,
		drifts: DriftEvent[]
	): Promise<void> {
		return this.sendDriftAlertTo(this.webhookUrl, projectName, result, drifts);
	}

	/**
	 * Posts a drift alert to a specific Discord webhook URL. An empty URL is a
	 * no-op — alerts are opt-in per project, so a user who doesn't configure
	 * Discord simply gets no alert, and no error is logged.
	 */
	async sendDriftAlertTo(
		webhookUrl: string,
		projectName: string,
		result: AnalysisResult | null,
		drifts: DriftEvent[]
	): Promise<void> {
		if (webhookUrl === '') return;
		if (result === null || result === undefined) {
			throw new Error('alerts: AnalysisResult is nil');
		}

		// RFC3339 with second precision, matching the original payload.
		const timestamp = new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');

		const embed: DiscordEmbed = {
			title: `⚠️ Drift Detected — ${projectName}`,
			color: colorForSeverity(result.severity),
			fields: [
				{ name: 'Severity', value: result.severity, inline: true },
				{
					name: 'Containers Affected',
					value: String(uniqueContainerCount(drifts)),
					inline: true
				},
				{ name: 'Summary', value: result.summary },
				{ name: 'Fix Command', value: '```bash\n' + result.fixCommand + '\n```' },
				{ name: 'Explanation', value: result.explanation }
			],
			footer: { text: `DriftWatch • detected at ${timestamp}` },
			timestamp
		};

		const payload: DiscordPayload = { embeds: [embed] };

		const res = await this.fetchImpl(webhookUrl, {
			method: 'POST',
			headers: { 'Content-Type': 'application/json' },
			body: JSON.stringify(payload),
			signal: AbortSignal.timeout(this.timeoutMs)
		});

		if (!res.ok) {
			const body = await res.text().catch(() => '');
			throw new Error(`discord webhook returned ${res.status}: ${body}`);
		}
	}

	/**
	 * Returns true if it's been at least 30 minutes since the last alert for
	 * this project (or there's no prior alert recorded).
	 *
	 * projectId is accepted for forward compatibility with per-project state,
	 * but the current implementation is stateless and relies on the caller to
	 * pass the last-alerted timestamp from the DB.
	 */
	shouldAlert(_projectId: string, lastAlertedAt: Date | null): boolean {
		if (lastAlertedAt === null || lastAlertedAt === undefined) return true;
		return Date.now() - lastAlertedAt.getTime() >= COOLDOWN_MS;
	}
}
